### PR TITLE
Tracklet Processor Displaced

### PR DESCRIPTION
**Pull request by hardikroutray**
#### PR description:

This PR introduces Tracklet Processor Displaced which combines the current displaced tracking functionality (Tracklet Engine Displaced, Triplet Engine and TrackletCalculatorDisplaced) into a single module

#### PR validation:

Efficiency checks with Displaced Mu samples show agreement between original and combined configuration. Presented in L1TkAlgo Meeting on [17 May](https://indico.cern.ch/event/1156404/#46-trackprocessordisplaced-mod). Also see introductory talk [here](https://indico.cern.ch/event/1117241/#34-status-of-displaced-track-i)

### CODE REVIEW

**Comment by tomalin** [comment]
Why is this PR changing the displaced tracklet wiring files in TrackFindingTracklet/data/ ? 

data/ files are not put in the CMSSW repo, instead being stored in https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git . They then automatically become available in $CMSSW_RELEASE_BASE/external/ since CMSSW_12_0_0_pre4 . During development, we usually put such data/ files in https://github.com/cms-L1TK/L1Trigger-TrackFindingTracklet.git .  But our git CI doesn't bother checking this out at the moment.

I suggest as a temporary work-around, you leave these data/ files where you have put them (if you do need to change them). This will allow git CI to pass. But then before we make a PR to central CMSSW, I'll delete them, and make a separate PR for them to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git  .

File: L1Trigger/TrackFindingTracklet/interface/Settings.h [hunk 1]
```diff
@@ -638,10 +638,15 @@ namespace trklet {
                                                                   {{6, 6, 6, 6, 10, 10, 10, 10, 0, 0, 6, 0}},
                                                                   {{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 6}}}};
 
+    /* std::array<std::array<unsigned int, N_SEED>, 3> lutwidthtabextended_{ */
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/interface/Settings.h:645]: Why is the lutwidthtabextended changed? And why can't we delete the commented out code?

File: L1Trigger/TrackFindingTracklet/interface/TrackletProcessorDisplaced.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+namespace trklet {
+
+  class Settings;
+  class Globals;
+  class MemoryBase;
+  class AllStubsMemory;
+  class AllInnerStubsMemory;
+  class VMStubsTEMemory;
+  class StubPairsMemory;
+
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/interface/TrackletProcessorDisplaced.h:26]: All CMSSW class header files must have a comment at the top explaining what they do. And a comment explaining any non-obvious member functions.

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+
+#include <utility>
+#include <tuple>
+
+using namespace std;
+using namespace trklet;
+
+TrackletProcessorDisplaced::TrackletProcessorDisplaced(string name, Settings const& settings, Globals* globals)
+    : TrackletCalculatorDisplaced(name, settings, globals),
+      // tebuffer_(CircularBuffer<TEData>(3), 0, 0, 0, 0),
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc]: Can we delete this commented out line?

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+  middleallstubs_.clear();
+  outerallstubs_.clear();
+  stubpairs_.clear();
+  innervmstubs_.clear();
+  outervmstubs_.clear();
+
+  // iAllStub_ = -1;
+  layerdisk_ = initLayerDisk(4);
+
+  unsigned int region = name.substr(1)[9] - 'A';
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc]: Magic numbers line "4" & "9" are not allowed. Replaced with named constants. I see several such constants in this class. Please eliminate them all.

**Comment by tomalin** [comment]
If not already done, please check performance OK on a 200PU sample, such as DisplacedMu + 200PU https://twiki.cern.ch/twiki/bin/view/CMS/L1TrackMC .

**Comment by hardikroutray** [comment]
> If not already done, please check performance OK on a 200PU sample, such as DisplacedMu + 200PU https://twiki.cern.ch/twiki/bin/view/CMS/L1TrackMC .

Performance consistent between original and combined configuration for 200PU sample. Efficiency checks on slides 16,17 and 18 [here](https://indico.cern.ch/event/1156404/#46-trackprocessordisplaced-mod)

**Comment by hardikroutray** [comment]
> Why is this PR changing the displaced tracklet wiring files in TrackFindingTracklet/data/ ?
> 
> data/ files are not put in the CMSSW repo, instead being stored in https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git . They then automatically become available in $CMSSW_RELEASE_BASE/external/ since CMSSW_12_0_0_pre4 . During development, we usually put such data/ files in https://github.com/cms-L1TK/L1Trigger-TrackFindingTracklet.git . But our git CI doesn't bother checking this out at the moment.
> 
> I suggest as a temporary work-around, you leave these data/ files where you have put them (if you do need to change them). This will allow git CI to pass. But then before we make a PR to central CMSSW, I'll delete them, and make a separate PR for them to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git .

Yes, data/ files need to be changed for the TPD module to work correctly. As you suggested, it would definitely be better if you could delete this and add these files to cms-data in a separate commit before PR to central CMSSW.

File: L1Trigger/TrackFindingTracklet/interface/Settings.h (hunk 1 above)
Comment (hardikroutray) [reply L1Trigger/TrackFindingTracklet/interface/Settings.h:645]: lutwidthtabextended is changed to take into account the fact that the maximum number of bits needed to represent lutval is not 0 for triplet seeds. The width still needs to be optimized before we settle on a final value. 21 was found to be a good maximum number to not face the error complaining of too many bits. (The actual widths should be smaller). I have kept the previous (probably optimized for previous version) values to use as a reference when further studies are performed on the widths.

**Comment by tomalin** [comment]
> > Why is this PR changing the displaced tracklet wiring files in TrackFindingTracklet/data/ ?
> > data/ files are not put in the CMSSW repo, instead being stored in https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git . They then automatically become available in $CMSSW_RELEASE_BASE/external/ since CMSSW_12_0_0_pre4 . During development, we usually put such data/ files in https://github.com/cms-L1TK/L1Trigger-TrackFindingTracklet.git . But our git CI doesn't bother checking this out at the moment.
> > I suggest as a temporary work-around, you leave these data/ files where you have put them (if you do need to change them). This will allow git CI to pass. But then before we make a PR to central CMSSW, I'll delete them, and make a separate PR for them to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git .
> 
> Yes, data/ files need to be changed for the TPD module to work correctly. As you suggested, it would definitely be better if you could delete this and add these files to cms-data in a separate commit before PR to central CMSSW.

Please explain how the files in data/ have been changed. I believe these files for the displaced  tracking are generated by the python scripts, such as HourGlassConfig.py that are documented in https://github.com/cms-L1TK/project_generation_scripts/blob/master/archive/README.md .  But there are no new PRs from you changing these python scripts.

**Comment by aehart** [comment]
> > > Why is this PR changing the displaced tracklet wiring files in TrackFindingTracklet/data/ ?
> > > data/ files are not put in the CMSSW repo, instead being stored in https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git . They then automatically become available in $CMSSW_RELEASE_BASE/external/ since CMSSW_12_0_0_pre4 . During development, we usually put such data/ files in https://github.com/cms-L1TK/L1Trigger-TrackFindingTracklet.git . But our git CI doesn't bother checking this out at the moment.
> > > I suggest as a temporary work-around, you leave these data/ files where you have put them (if you do need to change them). This will allow git CI to pass. But then before we make a PR to central CMSSW, I'll delete them, and make a separate PR for them to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git .
> > 
> > 
> > Yes, data/ files need to be changed for the TPD module to work correctly. As you suggested, it would definitely be better if you could delete this and add these files to cms-data in a separate commit before PR to central CMSSW.
> 
> Please explain how the files in data/ have been changed. I believe these files for the displaced tracking are generated by the python scripts, such as HourGlassConfig.py that are documented in https://github.com/cms-L1TK/project_generation_scripts/blob/master/archive/README.md . But there are no new PRs from you changing these python scripts.

@gershtein produced these files for us. I believe the process was basically to take the existing files for the extended tracking and replace each TCD, along with the TEDs and TREs that feed into it, with a TPD. Perhaps we can add any relevant scripts to project_generation_scripts next week when @gershtein gets back.

The ultimate plan is to update the TrackletConfigBuilder so that the wiring for the extended project can be generated dynamically, like that for the standard project is. But this will take a little longer.

**Comment by tomalin** [comment]
> > > > Why is this PR changing the displaced tracklet wiring files in TrackFindingTracklet/data/ ?
> > > > data/ files are not put in the CMSSW repo, instead being stored in https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git . They then automatically become available in $CMSSW_RELEASE_BASE/external/ since CMSSW_12_0_0_pre4 . During development, we usually put such data/ files in https://github.com/cms-L1TK/L1Trigger-TrackFindingTracklet.git . But our git CI doesn't bother checking this out at the moment.
> > > > I suggest as a temporary work-around, you leave these data/ files where you have put them (if you do need to change them). This will allow git CI to pass. But then before we make a PR to central CMSSW, I'll delete them, and make a separate PR for them to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git .
> > > 
> > > 
> > > Yes, data/ files need to be changed for the TPD module to work correctly. As you suggested, it would definitely be better if you could delete this and add these files to cms-data in a separate commit before PR to central CMSSW.
> > 
> > 
> > Please explain how the files in data/ have been changed. I believe these files for the displaced tracking are generated by the python scripts, such as HourGlassConfig.py that are documented in https://github.com/cms-L1TK/project_generation_scripts/blob/master/archive/README.md . But there are no new PRs from you changing these python scripts.
> 
> @gershtein produced these files for us. I believe the process was basically to take the existing files for the extended tracking and replace each TCD, along with the TEDs and TREs that feed into it, with a TPD. Perhaps we can add any relevant scripts to project_generation_scripts next week when @gershtein gets back.
> 
> The ultimate plan is to update the TrackletConfigBuilder so that the wiring for the extended project can be generated dynamically, like that for the standard project is. But this will take a little longer.

Understood. We shouldn't really on hand-edited wiring maps in the data/ directory, since otherwise we wouldn't be able to update or validate them. So we do need @gershtein to make a separate PR to the project_generation_scripts directory , such that the scripts there can produce the new data/ files. (And yes, add this functionality to TrackletConfigBuilder in the longer term ...)

File: L1Trigger/TrackFindingTracklet/interface/Settings.h (hunk 1 above)
Comment (tomalin) [reply L1Trigger/TrackFindingTracklet/interface/Settings.h:645]: Please add two comments. (1) Explain what lutwidthtab*_ in Settings.h represents; (2) Add "TO DO: tune lutwidthtabextended_ values".

Comment (hardikroutray) [reply L1Trigger/TrackFindingTracklet/interface/Settings.h:645]: Done.

File: L1Trigger/TrackFindingTracklet/interface/TrackletProcessorDisplaced.h (hunk 2 above)
Comment (hardikroutray) [reply L1Trigger/TrackFindingTracklet/interface/TrackletProcessorDisplaced.h:26]: added a comment on the top.

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc (hunk 3 above)
Comment (hardikroutray) [reply L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc]: Done

**Comment by tomalin** [comment]
Would the official CMSSW L1 track code be broken or damaged by your new TrackFindingTracklet/data/ files? i.e. Are your new data/ files reliant on any of the code change in this PR to keep the existing prompt & displaced tracking working OK?
I'd like to understand if you can make a PR for your data/ files now to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git , without us needing to wait for cms-L1TK to be merged into official CMSSW first?

**Comment by tomalin** [comment]
@hardikroutray This PR looks almost ready to merge, aside from the unanswered comment about magic numbers.

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc (hunk 4 above)
Comment (hardikroutray) [reply L1Trigger/TrackFindingTracklet/src/TrackletProcessorDisplaced.cc]: Done

**Comment by hardikroutray** [comment]
> @hardikroutray This PR looks almost ready to merge, aside from the unanswered comment about magic numbers.

have answered it now.

**Comment by hardikroutray** [comment]
> Would the official CMSSW L1 track code be broken or damaged by your new TrackFindingTracklet/data/ files? i.e. Are your new data/ files reliant on any of the code change in this PR to keep the existing prompt & displaced tracking working OK? I'd like to understand if you can make a PR for your data/ files now to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git , without us needing to wait for cms-L1TK to be merged into official CMSSW first?

The official CMSSW L1 track code shouldn't be affected by the files in TrackFindingTracklet/data/. Maybe @aehart can further comment on this.

**Review by tomalin — APPROVED** [review]
@gershtein @hardikroutray @aehart I'm approving this, but:
1) You need to make a PR to https://github.com/cms-data/L1Trigger-TrackFindingTracklet.git with your new data/ files for the displaced tracking.
2) You need to make a PR to https://github.com/cms-L1TK/project_generation_scripts/blob/master/archive/ to update the python scripts used to produce these data/ files . Otherwise they're likely to be overwritten next time some uses these scripts.

**Pull request merged**